# **Data Quality Check and Cleaning Pipeline**

Group TXB 13

## Table of Contents

- [1. Load In Data](#1-Load-In-Data)
- [2. Data Quality Check](#2-Data-Quality-Check)
  - [2.1 Dataset Overview and Data Types](#21-Dataset-Overview-and-Data-Types)
  - [2.2 Missing Values](#22-Missing-Values)
  - [2.3 Duplicates](#23-Duplicates)
  - [2.4 Numeric Columns Overview](#24-Numeric-Columns-Overview)
  - [2.5 Categorical Column Overview](#25-Categorical-Column-Overview)
  - [2.6 Date Check](#26-Date-Check)
  - [2.7 DQ AUDIT SUMMARY & CLEANING ROADMAP](#27-DQ-AUDIT-SUMMARY-CLEANING-ROADMAP)
- [3. Master cleaning pipeline](#3-Master-cleaning-pipeline)
  - [3.1 Income](#31-Income)
  - [3.2 Duplicates](#32-Duplicates)
  - [3.3 Date of birth](#33-Date-of-birth)
  - [3.4 Email validation](#34-Email-validation)
  - [3.5 Gender standardization](#35-Gender-standardization)
  - [3.6 Negative credit index fix](#36-Negative-credit-index-fix)
  - [3.7 Clean Data Summary](#37-Clean-Data-Summary)
  - [3.8 Save Clean Data as .json](#38-Save-Clean-Data-as-json)

## **1. Load In Data**

In [80]:
# 1. LOAD RAW DATA
import pandas as pd
import numpy as np
from pathlib import Path
import json
from datetime import datetime

data_path = Path("../data/raw_credit_applications.json")
with open(data_path, 'r') as f:
    raw = json.load(f)

df_raw = pd.json_normalize(raw)
print(f"Raw data: {len(df_raw)} rows x {len(df_raw.columns)} cols")
df_raw.head(2)

Raw data: 502 rows x 21 cols


,_id,spending_behavior,processing_timestamp,applicant_info.full_name,applicant_info.email,applicant_info.ssn,applicant_info.ip_address,applicant_info.gender,applicant_info.date_of_birth,applicant_info.zip_code,...,financials.credit_history_months,financials.debt_to_income,financials.savings_balance,decision.loan_approved,decision.rejection_reason,loan_purpose,decision.interest_rate,decision.approved_amount,financials.annual_salary,notes
0,app_200,"[{'category': 'Shopping', 'amount': 480}, {'ca...",2024-01-15T00:00:00Z,Jerry Smith,jerry.smith17@hotmail.com,596-64-4340,192.168.48.155,Male,2001-03-09,10036,...,23,0.20,31212,False,algorithm_risk_score,NaN,NaN,NaN,NaN,NaN
1,app_037,"[{'category': 'Rent', 'amount': 608}, {'catego...",NaN,Brandon Walker,brandon.walker2@yahoo.com,425-69-4784,10.1.102.112,M,1992-03-31,10032,...,51,0.18,17915,False,algorithm_risk_score,NaN,NaN,NaN,NaN,NaN


Successfully loaded nested JSON to flat DataFrame

Shape: 502 x 21 columns

Next: Data Quality Audit

## **2. Data Quality Check**

### 2.1 Dataset Overview and Data Types

In [81]:
# 2.1 DATASET OVERVIEW & DTYPES
print("SHAPE:", df_raw.shape)
print("\nDTYPES (sorted):")
print(df_raw.dtypes.sort_values())

SHAPE: (502, 21)

DTYPES (sorted):
decision.loan_approved                 bool
financials.savings_balance            int64
financials.credit_history_months      int64
decision.approved_amount            float64
decision.interest_rate              float64
financials.debt_to_income           float64
financials.annual_salary            float64
_id                                  object
loan_purpose                         object
decision.rejection_reason            object
financials.annual_income             object
applicant_info.date_of_birth         object
applicant_info.gender                object
applicant_info.ip_address            object
applicant_info.ssn                   object
applicant_info.email                 object
applicant_info.full_name             object
processing_timestamp                 object
spending_behavior                    object
applicant_info.zip_code              object
notes                                object
dtype: object


### 2.1 FINDINGS

**Dataset info:**
- Shape: **502 rows × 21 columns** Normal size

**Critical Data Type Issues (CONSISTENCY):**
- `financials.annual_income`: **object** (should be numeric)  
  *Cleaning:* `pd.to_numeric(errors='coerce')`

**Other observations:**
- 6 numeric columns OK (int64/float64)
- 15 object columns (some should be numeric/date)
- `spending_behavior`: object → likely list → handle separately

**DQ Impact:** High – blocks numeric analysis [file:3]

### 2.2 Missing Values

In [82]:
# 2.2 MISSING VALUES ANALYSIS
missing = df_raw.isna().sum().sort_values(ascending=False)
missing_pct = (missing / len(df_raw)) * 100
missing_df = pd.DataFrame({
    'missing_count': missing,
    'missing_pct': missing_pct.round(1)
}).query('missing_pct > 0')
print(missing_df.head(10))

                           missing_count  missing_pct
notes                                500         99.6
financials.annual_salary             497         99.0
loan_purpose                         452         90.0
processing_timestamp                 440         87.6
decision.rejection_reason            292         58.2
decision.interest_rate               210         41.8
decision.approved_amount             210         41.8
applicant_info.ssn                     5          1.0
applicant_info.ip_address              5          1.0
financials.annual_income               5          1.0


### 2.2 FINDINGS

**Top Missing (COMPLETENESS dimension):**

| Column                    | Missing Count | Missing % |
|---------------------------|---------------|-----------|
| notes                     | **500**       | **99.6%** |
| financials.annual_salary  | 497           | **99.0%** |
| loan_purpose              | 452           | **90.0%** |
| processing_timestamp      | 440           | **87.6%** |
| decision.rejection_reason | 292           | **58.2%** |
| decision.interest_rate    | 210           | **41.8%** |
| decision.approved_amount  | 210           | **41.8%** |
| applicant_info.ssn        | 5             | **1.0%**  |
| applicant_info.ip_address | 5             | **1.0%**  |
| financials.annual_income  | 5             | **1.0%**  |

**Business interpretation:**
- `rejection_reason`, `interest_rate`, `approved_amount`: expected missing (only for rejected/approved)
- `notes`, `annual_salary`, `loan_purpose`: **severe completeness issues**
  
**Cleaning steps:**
1. Drop `notes` (99.6% missing = useless)
2. Flag `annual_salary` as unreliable (99% missing)
3. Impute or drop high-missing rows for `loan_purpose`

**DQ Impact:** Medium-High – affects analysis reliability [file:3]


### 2.3 Duplicates

In [83]:
# 2.3 DUPLICATES CHECK (safe for list columns)
print("Checking duplicates (excluding list columns)...")

# Exclude unhashable columns (lists)
hashable_cols = []
for col in df_raw.columns:
    try:
        if df_raw[col].dropna().empty:
            hashable_cols.append(col)
        elif not any(isinstance(x, list) for x in df_raw[col].dropna().head()):
            hashable_cols.append(col)
    except:
        pass

df_check = df_raw[hashable_cols].copy()
dup_count = df_check.duplicated().sum()
dup_pct = round((dup_count / len(df_check)) * 100, 2)

print(f"Duplicate rows: {dup_count} ({dup_pct}%)")
if dup_count > 0:
    print("\nSample duplicates:")
    print(df_check[df_check.duplicated()].head(2))
else:
    print("No duplicates found")

Checking duplicates (excluding list columns)...
Duplicate rows: 0 (0.0%)
No duplicates found


### 2.3 FINDINGS
**Results:**
- Duplicate rows: **0 (0.0%)** -> **Clean!**

**Notes:**
- Checked excluding unhashable list columns (`spending_behavior`)
- No exact row duplicates found

**DQ dimension:** ACCURACY -> **No action needed**

**Business key check recommendation:** 
- Verify `_id` uniqueness: `df_raw['_id'].nunique() == len(df_raw)`


### 2.4 Numeric Columns Overview

In [84]:
# 2.4 NUMERIC COLUMNS OVERVIEW

num_cols = [
    'financials.credit_history_months', 
    'financials.debt_to_income', 
    'financials.savings_balance', 
    'decision.interest_rate', 
    'decision.approved_amount', 
    'financials.annual_salary'
]

temp_income = pd.to_numeric(df_raw['financials.annual_income'], errors='coerce')

print("\nNUMERIC SUMMARY")
print("-" * 60)
print(f"{'Column':<25} {'Min':>8} {'Max':>8} {'Mean':>8} {'Miss%':>6} {'Negatives'}")

# Print special row for Annual Income
print(f"{'financials.annual_income':<25} {temp_income.min():>8.1f} {temp_income.max():>8.1f} {temp_income.mean():>8.1f} {temp_income.isna().mean()*100:>5.1f}% 0")

for col in num_cols:
    stats = df_raw[col].describe()
    min_val = stats['min']
    max_val = stats['max']
    mean_val = stats['mean']
    miss_pct = round(df_raw[col].isna().mean()*100, 1)
    negatives = (df_raw[col] < 0).sum()
    
    neg_flag = f"{negatives} ({negatives/len(df_raw)*100:.1f}%)" if negatives > 0 else "0"
    
    print(f"{col:<25} {min_val:>8.1f} {max_val:>8.1f} {mean_val:>8.1f} {miss_pct:>6.1f}% {neg_flag}")


NUMERIC SUMMARY
------------------------------------------------------------
Column                         Min      Max     Mean  Miss% Negatives
financials.annual_income       0.0 171000.0  82705.1   1.0% 0
financials.credit_history_months    -10.0    133.0     50.4    0.0% 2 (0.4%)
financials.debt_to_income      0.1      1.9      0.2    0.0% 0
financials.savings_balance  -5000.0  88078.0  29493.5    0.0% 1 (0.2%)
decision.interest_rate         2.5      6.5      4.6   41.8% 0
decision.approved_amount   15000.0  80000.0  47845.9   41.8% 0
financials.annual_salary   45000.0  94000.0  69200.0   99.0% 0


In [85]:
#2.4 Logical Check: Debt-to-Income Ratio
col_dti = 'financials.debt_to_income'

print("LOGIC CHECK: DEBT-TO-INCOME")
print("-" * 35)
print(f"Average DTI: {df_raw[col_dti].mean():.2f}")
print(f"Max DTI found: {df_raw[col_dti].max():.2f}")

high_dti_mask = df_raw[col_dti] > 1.0
high_dti_count = high_dti_mask.sum()

print(f"Records where Debt > Income (DTI > 1.0): {high_dti_count}")

if high_dti_count > 0:
    print("\nSample of High DTI Records:")
    display(df_raw[high_dti_mask][[col_dti, 'financials.annual_income']].head())

LOGIC CHECK: DEBT-TO-INCOME
-----------------------------------
Average DTI: 0.25
Max DTI found: 1.85
Records where Debt > Income (DTI > 1.0): 1

Sample of High DTI Records:


,financials.debt_to_income,financials.annual_income
316,1.85,88000


### 2.4 FINDINGS

**Key Observations:**
- **6 numeric columns detected** (vs 7 expected) → `annual_income` object as is
- Range looks reasonable except for specific logical outliers in the debt ratio.

**Critical Issues:**
1. **`financials.credit_history_months`**: **Min < 0** (2 records, 0.4%) -> **Invalid**
   - *Cleaning:* `clip(lower=0)` or `set to NaN`
   
2. **`financials.savings_balance`**: **Min < 0** (1 record, 0.2%) -> **Borderline (overdraft)**
   - *Cleaning:* Keep as business valid but flag for audit.

3. **`financials.debt_to_income`**: **Max > 1.0** (Found records up to 1.9) -> **Logical Outlier**
   - *Governance Note:* Several records show debt exceeding annual income. While possibly valid for high-risk profiles, these must be flagged for the bias analysis phase.

**Cleaning steps summary:**
- Convert `annual_income` to numeric (float).
- `df['credit_history_months'] = df['credit_history_months'].clip(lower=0)`
- Create `is_overdrawn` flag for negative savings balances.
- Create `high_dti_flag` for records where the debt ratio > 1.0.

### 2.5 Categorical Column Overview

In [86]:
# 2.5 CATEGORICAL OVERVIEW (safe)
print("ALL CATEGORICAL - Uniques & Top values")
print("="*60)

for col in df_raw.columns:
    if df_raw[col].dtype == 'object':
        print(f"\n{col}:")
        try:
            vc = df_raw[col].value_counts(dropna=False).head(4)
            print(vc)
        except:
            print("  [SKIPPED - list column]")

ALL CATEGORICAL - Uniques & Top values

_id:
_id
app_001    2
app_042    2
app_453    1
app_177    1
Name: count, dtype: int64

spending_behavior:
spending_behavior
[{'category': 'Insurance', 'amount': 171}]    2
[{'category': 'Utilities', 'amount': 566}]    2
[{'category': 'Utilities', 'amount': 786}]    2
[{'category': 'Fitness', 'amount': 98}]       2
Name: count, dtype: int64

processing_timestamp:
processing_timestamp
NaN                     440
2024-01-15T00:00:00Z     59
2025-12-01T00:00:00Z      1
2026-03-15T00:00:00Z      1
Name: count, dtype: int64

applicant_info.full_name:
applicant_info.full_name
Susan Flores    3
Rachel King     2
George Clark    2
James Rivera    2
Name: count, dtype: int64

applicant_info.email:
applicant_info.email
                               7
joseph.lopez1@gmail.com        2
stephanie.nguyen47@mail.com    2
nicholas.king46@outlook.com    1
Name: count, dtype: int64

applicant_info.ssn:
applicant_info.ssn
NaN            5
780-24-9300    2
652-70-55

### 2.5 FINDINGS

**Top Issues:**

1. **`applicant_info.gender`**: **Severe mixed coding**  
   - Male: 195, Female: 193, **F: 58**, *(M is likely present)*  
   - *Issue:* 3-4+ formats → unusable for bias analysis  
   - *Fix:* `replace({'F': 'Female', 'M': 'Male'}).str.capitalize()`

2. **`_id`**: **PRIMARY KEY VIOLATION**   
   - app_001: **2x**, app_042: **2x**  
   - *Issue:* Duplicated application IDs  
   - *Fix:* `drop_duplicates(subset='_id', keep='first')`

3. **`applicant_info.full_name`**: **Person-level duplicates**   
   - Susan Flores: **3x** occurrences  
   - *Issue:* Same individual multiple apps?  

4. **`financials.annual_income`**: **String numbers** (2.1 confirm)  
   - Top: "79000"(11x), "100000"(10x)  
   - *Fix:* `pd.to_numeric(errors='coerce')`

5. **`applicant_info.email`**: **Empty PII**  
   - **7 completely blank** → privacy/governance flag

**Priority Summary:**
| Priority | Column | Issue Type | Fix |
|----------|--------|------------|-----|
| **CRITICAL** | _id | Duplicate PK | drop_duplicates() |
| **HIGH** | gender | Mixed coding | replace mapping |
| MEDIUM | annual_income | Type mismatch | to_numeric() |
| LOW | full_name | Potential duplicates | business decision |

**Business Impact:** Gender unusable for **bias analysis**; trust issues with IDs

### 2.6 Date Check


In [87]:
# 2.6 DATE CONSISTNECY FORMAT CHECK - Formats
date_col = 'applicant_info.date_of_birth'

print("DATE OF BIRTH FORMATS ANALYSIS")
print("-" * 50)

# Sample non-null dates
dates = df_raw[date_col].dropna().unique()[:20]
print(f"Unique date formats (sample {len(dates)}):")
for d in dates:
    print(f"  '{d}' → len: {len(str(d))}, format: {pd.to_datetime(d, errors='coerce', infer_datetime_format=True)}")

# Format patterns
date_patterns = df_raw[date_col].dropna().str.len().value_counts().sort_index()
print(f"\nDate string lengths: {dict(date_patterns)}")

# Try parsing
df_raw['date_parsed'] = pd.to_datetime(df_raw[date_col], errors='coerce')
invalid_dates = df_raw['date_parsed'].isna().sum()
print(f"\nInvalid dates: {invalid_dates} ({invalid_dates/len(df_raw)*100:.1f}%)")

DATE OF BIRTH FORMATS ANALYSIS
--------------------------------------------------
Unique date formats (sample 20):
  '2001-03-09' → len: 10, format: 2001-03-09 00:00:00
  '1992-03-31' → len: 10, format: 1992-03-31 00:00:00
  '1989-10-24' → len: 10, format: 1989-10-24 00:00:00
  '1983-04-25' → len: 10, format: 1983-04-25 00:00:00
  '1999-05-21' → len: 10, format: 1999-05-21 00:00:00
  '14/02/1982' → len: 10, format: 1982-02-14 00:00:00
  '28/01/1990' → len: 10, format: 1990-01-28 00:00:00
  '1991-10-11' → len: 10, format: 1991-10-11 00:00:00
  '1990-05-04' → len: 10, format: 1990-05-04 00:00:00
  '1989-10-10' → len: 10, format: 1989-10-10 00:00:00
  '1970-10-01' → len: 10, format: 1970-10-01 00:00:00
  '01/12/1978' → len: 10, format: 1978-01-12 00:00:00
  '1989-06-13' → len: 10, format: 1989-06-13 00:00:00
  '1976-04-19' → len: 10, format: 1976-04-19 00:00:00
  '1990/07/26' → len: 10, format: 1990-07-26 00:00:00
  '1992-06-22' → len: 10, format: 1992-06-22 00:00:00
  '1997-03-23' → len:

C:\Users\mikis\AppData\Local\Temp\ipykernel_30120\949907516.py:11: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  print(f"  '{d}' → len: {len(str(d))}, format: {pd.to_datetime(d, errors='coerce', infer_datetime_format=True)}")
C:\Users\mikis\AppData\Local\Temp\ipykernel_30120\949907516.py:11: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  print(f"  '{d}' → len: {len(str(d))}, format: {pd.to_datetime(d, errors='coerce', infer_datetime_format=True)}")


In [88]:
#2.6 DATE CONSISTENCY CHECK: Future Timestamps

current_time = datetime.now()

df_raw['temp_ts'] = pd.to_datetime(df_raw['processing_timestamp'], errors='coerce').dt.tz_localize(None)
future_dates = df_raw[df_raw['temp_ts'] > current_time]

print("TEMPORAL VALIDITY CHECK")
print("-" * 30)
print(f"Records with Future Timestamps: {len(future_dates)}")

if len(future_dates) > 0:
    print(f"Note: Max date found in data is {df_raw['temp_ts'].max()}")
    print("\nSample of future-dated records:")
    display(df_raw[df_raw['temp_ts'] > current_time][['processing_timestamp']].head())

df_raw.drop(columns=['temp_ts'], inplace=True)

TEMPORAL VALIDITY CHECK
------------------------------
Records with Future Timestamps: 2
Note: Max date found in data is 2027-01-20 00:00:00

Sample of future-dated records:


,processing_timestamp
169,2026-03-15T00:00:00Z
441,2027-01-20T00:00:00Z


In [89]:
# 2.6 DATE CHECK - Blank/Empty DOBs
dob_col = 'applicant_info.date_of_birth'

blank_dob_mask = (
    df_raw[dob_col].isna() | 
    (df_raw[dob_col].astype(str).str.strip() == "")
)

blank_dob_count = blank_dob_mask.sum()

print("COMPLETENESS CHECK: DATE OF BIRTH")
print("-" * 40)
print(f"Missing/Blank DOBs: {blank_dob_count}")

if blank_dob_count > 0:
    print(f"Governance Action: These {blank_dob_count} records must be dropped as Age cannot be calculated.")

COMPLETENESS CHECK: DATE OF BIRTH
----------------------------------------
Missing/Blank DOBs: 5
Governance Action: These 5 records must be dropped as Age cannot be calculated.


### 2.6 FINDINGS

**Key Stats:**
- **497 unique date entries** (len=10), **4 completely empty** → **32.3% initial parse failure**.
- **Temporal Anomaly:** Identified **2 records** with processing timestamps set in the future (e.g., 2025/2026), violating temporal validity.

**Mixed Formats:**
| Format | Example | Parses? |
|--------|---------|---------|
| yyyy-mm-dd | 2001-03-09 | OK |
| mm/dd/yyyy | 03/20/1968 | OK (custom parser) |
| dd/mm/yyyy | 14/02/1982 | OK (custom parser) |
| yyyy/mm/dd | 1990/07/26 | OK (custom parser) |

**Problem:** **4 international formats + Future Dates** → Inconsistent parsing and logical errors in system metadata.

**DQ Category:** **CONSISTENCY & VALIDITY** (Critical for age-based bias analysis).

**Fix Plan:**
- **Future Dates:** Filter out records where `processing_timestamp` > `current_date` to ensure logical integrity.
- **Custom Parser:** Implement `parse_custom_date()` (Section 3.3) to resolve multi-separator logic and dd/mm vs mm/dd ambiguity.
- **Completeness:** Drop records with blank DOBs, as age is a mandatory protected characteristic for the bias audit.

**Business impact:** Logical errors and missing age data block the mandatory **Regulatory Bias Analysis** and violate **EU AI Act** data quality standards.

### 2.7 DQ AUDIT SUMMARY & CLEANING ROADMAP

**Issues sorted: Severity ↓ (6/6 dimensions covered!)**

| # | **Issue** | **Column** | **Chapter** | **DQ Category** | **Severity** | **Fix** |
|---|-----------|------------|-------------|-----------------|--------------|---------|
| 1 | String income | `financials.annual_income` | **2.1 Dtypes** | Accuracy | **HIGH** | `pd.to_numeric(errors='coerce')` |
| 2 | Duplicate IDs | `_id` | **2.3 Duplicates** | Uniqueness | **HIGH** | `drop_duplicates()` |
| 3 | Future Timestamps | `processing_timestamp` | **2.6 Dates** | **Validity** | **HIGH** | Filter out records > Today |
| 4 | Blank DOBs | `applicant_info.date_of_birth` | **2.6 Dates** | **Completeness**| **HIGH** | Drop records (Required for Bias Analysis) |
| 5 | 32% parse fail dates | `applicant_info.date_of_birth` | **2.6 Dates** | Validity | **HIGH** | `pd.to_datetime(dayfirst=True)` |
| 6 | 7 empty emails | `applicant_info.email` | **2.5 Categorical** | Validity | **MEDIUM** | Regex + flag invalid |
| 7 | High Debt (>1.0) | `financials.debt_to_income` | **2.4 Numeric** | **Consistency** | **MEDIUM** | Flag as `high_dti` |
| 8 | Gender `Male/F...` | `applicant_info.gender` | **2.5 Categorical** | Consistency | **MEDIUM** | `.str.lower()` + mapping |
| 9 | Negative durations | `financials.credit_history_months` | **2.4 Numeric** | Consistency | **LOW** | `.clip(lower=0)` |
| 10| Negative savings | `financials.savings_balance` | **2.4 Numeric** | **Validity** | **LOW** | Flag as `is_overdrawn` |

---

#### Governance Note: Achieving 6/6 DQ Dimensions
To ensure our pipeline meets the standards of a **"High-Risk AI System"** under the EU AI Act, we have addressed all six core dimensions of data quality:
1. **Accuracy:** Corrected data types for financial calculations.
2. **Completeness:** Identified and handled missing critical fields like Date of Birth.
3. **Consistency:** Standardized gender categories and handled logical time durations.
4. **Validity:** Flagged future timestamps and incorrectly formatted dates.
5. **Uniqueness:** Removed duplicate application IDs.
6. **Integrity:** Validated logical relationships between Debt and Income.

## **3. Master cleaning pipeline**

### 3.1 Income


In [90]:
# 3.1 INCOME [2.1 Dtypes] HIGH!
df1 = df_raw.copy()
df1['financials.annual_income'] = pd.to_numeric(df1['financials.annual_income'], errors='coerce')

failed_conversions = df1['financials.annual_income'].isna().sum() - df_raw['financials.annual_income'].isna().sum()

print(f"1. Income numeric conversion complete (dtype: {df1['financials.annual_income'].dtype})")
print(f"Successfully converted numeric strings.")
print(f"Warning: {failed_conversions} non-numeric strings were coerced to NaN for safety.")

1. Income numeric conversion complete (dtype: float64)
Successfully converted numeric strings.


### 3.2 Duplicates

In [91]:
# 3.2 DUPLICATES [2.5 Categorical] HIGH!
initial_count = len(df1)

df2 = df1.drop_duplicates(subset='_id', keep='first').copy()

duplicate_applicants = df2.duplicated(subset=['applicant_info.ssn'], keep=False).sum()

print(f"2. Uniqueness audit complete:")
print(f"Dropped {initial_count - len(df2)} duplicate Application IDs.")
print(f"Remaining rows: {len(df2)}")
print(f"Governance Alert: Found {duplicate_applicants} records sharing an SSN (potential multiple applications).")

print("Remaining _id dups:", df2['_id'].duplicated().sum())

2. Uniqueness audit complete:
Dropped 2 duplicate Application IDs.
Remaining rows: 500
Governance Alert: Found 8 records sharing an SSN (potential multiple applications).
Remaining _id dups: 0


### 3.3 Future Timestamp

In [92]:
# 3.3 FUTURE DATES [2.6 Date] HIGH!
current_time = datetime.now()
initial_count = len(df2)

df2_dates = df2.copy()
df2_dates['temp_ts'] = pd.to_datetime(df2_dates['processing_timestamp'], errors='coerce').dt.tz_localize(None)

is_future = df2_dates['temp_ts'] > current_time
df3_pre = df2_dates[~is_future].copy() 

print(f"3.3 Future date filter complete:")
print(f"Removed {is_future.sum()} records with confirmed future timestamps.")
print(f"Remaining records: {len(df3_pre)}")

# Cleanup
df3_pre = df3_pre.drop(columns=['temp_ts'])

3.3 Future date filter complete:
Removed 2 records with confirmed future timestamps.
Remaining records: 498


### 3.4 Blank Date of Birth

In [93]:
# 3.4 BLANK DOBs [2.6 Date] HIGH!
initial_count = len(df3_pre)

df3_filtered = df3_pre.dropna(subset=['applicant_info.date_of_birth']).copy()
df3_filtered = df3_filtered[df3_filtered['applicant_info.date_of_birth'].astype(str).str.strip() != ""].copy()

dropped_count = initial_count - len(df3_filtered)

print(f"3.4 Blank DOB filter complete:")
print(f"Dropped {dropped_count} records missing a valid Date of Birth.")
print(f"Records remaining for master pipeline: {len(df3_filtered)}")

3.4 Blank DOB filter complete:
Dropped 4 records missing a valid Date of Birth.
Records remaining for master pipeline: 494


### 3.5 Date of Birth Format

In [94]:
# 3.5 CUSTOM DATE PARSER [2.6 Date] HIGH!

def parse_custom_date(date_str):
    if pd.isna(date_str):
        return pd.NaT
    s = str(date_str).strip()
    if len(s) != 10:
        return pd.NaT
    try:
        if '/' in s:
            p = s.split('/')
            a, b, c = int(p[0]), int(p[1]), int(p[2])
            if b > 12: m, d, y = a, b, c  # Middle number > 12 means DD/MM/YYYY
            else: d, m, y = a, b, c       # Standard MM/DD/YYYY
        elif '-' in s:
            y, m, d = map(int, s.split('-'))
        elif int(s[2]) <= 3:
            d, m, y = int(s[0:2]), int(s[2:4]), int(s[4:8])
        else:
            y, m, d = int(s[0:4]), int(s[4:6]), int(s[6:8])
        return pd.to_datetime(f"{y}-{m:02d}-{d:02d}")
    except:
        return pd.NaT

df3 = df3_filtered.copy()

df3['parsed_dob'] = df3['applicant_info.date_of_birth'].apply(parse_custom_date)
df3['demographics.age'] = ((pd.Timestamp.now() - df3['parsed_dob']) // pd.Timedelta(days=365.25)).astype('Int64')

df3['demographics.age'] = df3['demographics.age'].clip(18, 110)

print(f"5. Date parsing and Age derivation complete:")
print(f"Usable Data: {df3['demographics.age'].notna().sum()} out of {len(df3)} records are now clean and age-verified.")
print(f"Age range: {df3['demographics.age'].min()} to {df3['demographics.age'].max()} years.")


C:\Users\mikis\AppData\Local\Temp\ipykernel_30120\4254841982.py:21: UserWarning: Parsing dates in %d-%m-%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  return pd.to_datetime(f"{y}-{m:02d}-{d:02d}")


5. Date parsing and Age derivation complete:
Usable Data: 494 out of 494 records are now clean and age-verified.
Age range: 23 to 67 years.


In [ ]:
# 4.1 CLEAN DATA SUMMARY
print("=" * 50)
print("CLEAN DATA SUMMARY")
print("=" * 50)
print(f"Shape: {df_clean.shape}")
print(f"Uniqueness: 0 duplicate Application IDs")
print(f"Accuracy: Income converted to {df_clean['financials.annual_income'].dtype}")

age_complete = df_clean['demographics.age'].notna().sum()
print(f"Completeness: {age_complete}/{len(df_clean)} records have valid Age")

print(f"Validity: {(~df_clean['email_valid']).sum()} malformed emails flagged")
print(f"Consistency: {df_clean['applicant_info.gender'].nunique()} gender categories")

print(f"Integrity: No negative credit months (min={df_clean['financials.credit_history_months'].min()})")
print(f"Integrity: No negative savings balances (min={df_clean['financials.savings_balance'].min()})")

CLEAN DATA SUMMARY
Shape: (493, 26)
Uniqueness: 0 duplicate Application IDs
Accuracy: Income converted to float64
Completeness: 493/493 records have valid Age
Validity: 7 malformed emails flagged
Consistency: 2 gender categories
Integrity: No negative credit months (min=0)
Integrity: No negative savings balances (min=0)


### 3.6 Email Validation

In [95]:
# 3.3 FUTURE DATES - Dimension: Validity
df4 = df3.copy()

email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
df4['email_valid'] = df4['applicant_info.email'].str.match(email_pattern, na=False)

valid_emails = df4['email_valid'].sum()
total_current = len(df4)
invalid_emails = (~df4['email_valid']).sum()

print(f"6. Email validation audit complete:")
print(f"Valid emails: {valid_emails} / {total_current}")
print(f"Flagged {invalid_emails} invalid or missing emails.")

if invalid_emails > 0:
    print(f"   Sample invalid: {df4[~df4['email_valid']]['applicant_info.email'].head(11).tolist()}")

6. Email validation audit complete:
Valid emails: 487 / 494
Flagged 7 invalid or missing emails.
   Sample invalid: ['mike johnson@gmail.com', 'test.user.outlook.com', '', 'john.doe@invalid', '', '', 'sarah.smith@']


### 3.7 High Debt

In [96]:
# 3.7 HIGH DEBT-TO-INCOME FLAG [2.4 Logical] HIGH!

df5 = df4.copy()

dti_threshold = 1.0
df5['high_dti_flag'] = df5['financials.debt_to_income'] > dti_threshold

high_dti_count = df5['high_dti_flag'].sum()
avg_dti = df5['financials.debt_to_income'].mean()

print(f"7. Debt-to-Income Logic Check complete:")
print(f"Average DTI in cleaned dataset: {avg_dti:.2f}")
print(f"Flagged {high_dti_count} records with DTI > {dti_threshold} (High Integrity Risk).")

# Quick check on the most extreme case
if high_dti_count > 0:
    max_dti = df5['financials.debt_to_income'].max()
    print(f"Note: Maximum DTI found is {max_dti:.2f}")

7. Debt-to-Income Logic Check complete:
Average DTI in cleaned dataset: 0.25
Flagged 1 records with DTI > 1.0 (High Integrity Risk).
Note: Maximum DTI found is 1.85


### 3.8 Gender Standardization

In [97]:
# 3.8 GENDER STANDARDIZE [2.5 Categorical] MEDIUM!
df6 = df5.copy()

df6['applicant_info.gender'] = (df6['applicant_info.gender']
    .str.lower()
    .replace({'m': 'male', 'f': 'female', '': None})
)

gender_counts = df6['applicant_info.gender'].value_counts().to_dict()
missing_gender = df6['applicant_info.gender'].isna().sum()
unique_genders = df6['applicant_info.gender'].nunique()

print(f"8. Gender standardization complete:")
print(f"Unique categories: {unique_genders} (Target: 2)")
print(f"Distribution: {gender_counts}")
print(f"Missing Gender: {missing_gender} records (Complexity for Bias Audit)")

8. Gender standardization complete:
Unique categories: 2 (Target: 2)
Distribution: {'female': 249, 'male': 244}
Missing Gender: 1 records (Complexity for Bias Audit)


### 3.9 Negative Duration

In [98]:
# 3.9 NEGATIVE DURATION [2.4 Numeric] LOW!
df7 = df6.copy()

before = (df7['financials.credit_history_months'] < 0).sum()
df7['financials.credit_history_months'] = df7['financials.credit_history_months'].clip(lower=0)
after = (df7['financials.credit_history_months'] < 0).sum()

print(f"9. Negative durations audit complete:")
print(f"Logical errors fixed: {before} → {after}")
print(f"Minimum credit history: {df7['financials.credit_history_months'].min()} months")


df_clean = df7

9. Negative durations audit complete:
Logical errors fixed: 2 → 0
Minimum credit history: 0 months


### 3.10 NEGATIVE SAVINGS

In [99]:
# 3.10 NEGATIVE SAVINGS [2.4 Numeric] LOW!

initial_count = len(df_clean)

df_clean = df_clean[df_clean['financials.savings_balance'] >= 0].copy()

dropped_count = initial_count - len(df_clean)

print(f"10. Negative savings audit complete:")
print(f"Removed {dropped_count} records with negative balances (Validity check).")
print(f"Final Minimum Savings: {df_clean['financials.savings_balance'].min()}")

print(f"\nMASTER PIPELINE FINALIZED")
print(f"Final Dataset Size: {df_clean.shape}")

10. Negative savings audit complete:
Removed 1 records with negative balances (Validity check).
Final Minimum Savings: 0

MASTER PIPELINE FINALIZED
Final Dataset Size: (493, 26)


## **4 Generate Data**

### 4.1 Clean Data Summary

In [100]:
# 4.1 CLEAN DATA SUMMARY
print("=" * 50)
print("CLEAN DATA SUMMARY")
print("=" * 50)
print(f"Shape: {df_clean.shape}")
print(f"Uniqueness: 0 duplicate Application IDs")
print(f"Accuracy: Income converted to {df_clean['financials.annual_income'].dtype}")

age_complete = df_clean['demographics.age'].notna().sum()
print(f"Completeness: {age_complete}/{len(df_clean)} records have valid Age")

print(f"Validity: {(~df_clean['email_valid']).sum()} malformed emails flagged")
print(f"Consistency: {df_clean['applicant_info.gender'].nunique()} gender categories")

print(f"Integrity: No negative credit months (min={df_clean['financials.credit_history_months'].min()})")
print(f"Integrity: No negative savings balances (min={df_clean['financials.savings_balance'].min()})")

CLEAN DATA SUMMARY
Shape: (493, 26)
Uniqueness: 0 duplicate Application IDs
Accuracy: Income converted to float64
Completeness: 493/493 records have valid Age
Validity: 7 malformed emails flagged
Consistency: 2 gender categories
Integrity: No negative credit months (min=0)
Integrity: No negative savings balances (min=0)


### 4.2 Save Clean Data as .json

In [101]:
# 4.2 SAVE CLEAN DATA - Final Handover

output_path = Path("../data/clean_credit_applications.json")

df_clean.to_json(output_path, orient='records', indent=2)

print(f"Success! Saved cleaned data to: {output_path}")
print(f"Final Audit Volume: {len(df_clean)} records")

Success! Saved cleaned data to: ..\data\clean_credit_applications.json
Final Audit Volume: 493 records
